net: added module for filtering packets based on l2 addresses

haukepetersen opened the pull request:
This module allows to filter network packets base on linker layer addresses. It can be used in two configurations, `blacklisting` or `whitelisting` packets. To chose a mode, simply include the right module:
```
USEMODULE=l2filter_blacklist make ...
```
or 
```
USEMODULE=l2filter_whitelist make ...
```

The actual filter list are then controlled via the `ifconfig` shell command:
```
> ifconfig 7 set l2filter 79:65:08:22:86:93:9d:5a
...
> ifconfig
2017-05-16 16:41:54,720 - INFO #  ifconfig
2017-05-16 16:41:54,724 - INFO # Iface  7   HWaddr: 23:2a  Channel: 26  NID: 0x23
2017-05-16 16:41:54,729 - INFO #            Long HWaddr: 0e:46:6e:50:7d:d4:23:2a 
2017-05-16 16:41:54,732 - INFO #            TX-Power: 0dBm  State: IDLE 
2017-05-16 16:41:54,738 - INFO #            ACK_REQ  AUTOCCA  MTU:1280  HL:64  6LO  RTR  IPHC  
2017-05-16 16:41:54,741 - INFO #            Source address length: 8
2017-05-16 16:41:54,743 - INFO #            Link type: wireless
2017-05-16 16:41:54,749 - INFO #            inet6 addr: ff02::1/128  scope: local [multicast]
2017-05-16 16:41:54,755 - INFO #            inet6 addr: fe80::c46:6e50:7dd4:232a/64  scope: local
2017-05-16 16:41:54,761 - INFO #            inet6 addr: ff02::1:ffd4:232a/128  scope: local [multicast]
2017-05-16 16:41:54,766 - INFO #            inet6 addr: ff02::1a/128  scope: local [multicast]
2017-05-16 16:41:54,767 - INFO #            
2017-05-16 16:41:54,771 - INFO #            Black-listed link layer addresses:
2017-05-16 16:41:54,775 - INFO #              0: 79:65:08:22:86:93:9d:5a
2017-05-16 16:41:54,775 - INFO # 
2017-05-16 16:41:54,778 - INFO #            Statistics for Layer 2
2017-05-16 16:41:54,781 - INFO #             RX packets 10  bytes 394
2017-05-16 16:41:54,786 - INFO #             TX packets 12 (Multicast: 6)  bytes 431
2017-05-16 16:41:54,789 - INFO #             TX succeeded 11 errors 1
2017-05-16 16:41:54,791 - INFO #            Statistics for IPv6
2017-05-16 16:41:54,794 - INFO #             RX packets 10  bytes 554
2017-05-16 16:41:54,799 - INFO #             TX packets 12 (Multicast: 6)  bytes 682
2017-05-16 16:41:54,802 - INFO #             TX succeeded 12 errors 0
2017-05-16 16:41:54,803 - INFO # 
```

So far this module is hooked for all Ethernet and IEEE802.15.4 devices.

Missing/open issues (will get attention in follow-up PRs):
- so far addresses can not be removed via `ifconfig` -> @PeterKietzmann will add this
- a way to pass static list during compile time would be nice
- currently the `netstats` module counts packets that are dropped by the filter -> PR will follow soon

miri64 commented on sys/include/net/netopt.h:
```diff
@@ -248,6 +248,11 @@ typedef enum {
     NETOPT_NUMOF,
+
+    /**
+     * @brief   add/remove an address to the link layer filter list
```
What are the required types? How are removal and addition distinguished?

PeterKietzmann replied:
- "remove" should be removed here as it needs an other netopt parameter. Addition and removal are not distinguished here
- expects a pointer to `l2filter_t struct` on`get` or to a hardware address (usually `uint8_t`) on `set`

PeterKietzmann commented on sys/net/link_layer/l2filter/l2filter.c:
```diff
@@ -0,0 +1,106 @@
+    }
+}
+
+int l2filter_add(l2filter_t *list, const void *addr, size_t addr_len)
```
This function does not check for duplicate entries. Comparing the IPv6 blacklist module they used a bitfield to detect existing entries, for performance reasons. I can add check later, if you don't have time. @haukepetersen, @miri64 what's you opinion. Should we rather
1. not check for duplicates
2.  iterate through the list (via '`match()` function)
3. add a bitfield as done in IPv6 blacklist
IMO we should give 2. a try.

miri64 replied:
TBH I don't have a strong opinion on either option. Since this is a user operation 1. is the most efficient if documented well but it would be different to the IPv6 counterpart indeed. Also I'm not sure which implications this would have for removal.

haukepetersen replied:
I decided for (1) intentionally. IMHO this feature is basically made for testing and experiments, where in both cases one has pretty good control and knowledge what one is doing. So there is IMHO no need to catch duplicate entries...

PeterKietzmann commented:
I can confirm both blacklist and whitelist working on native and some boards. IMO this is a very useful tool which I will need in very near future. However, I'm not 100% happy with introducing two new `NETOPT` types to add and remove (will be done in a follow-up) entries. @miri64 what's your general opinion on this?
BTW: I will approve after minor comments are addressed and there is general consensus about the concept.

PeterKietzmann commented:
Please have a look at [#38](https://github.com/haukepetersen/RIOT/pull/38)

miri64 commented on sys/include/net/netopt.h:
```diff
@@ -248,6 +248,16 @@ typedef enum {
     NETOPT_NUMOF,
+
+    /**
+     * @brief   add/remove an address to/from the link layer filter list
```
Isn't `NETOPT_L2FILTER_RM` for remove now?

miri64 commented on sys/include/net/netopt.h:
```diff
@@ -248,6 +248,16 @@ typedef enum {
+     * blacklisting and whitelisting.
+     */
+    NETOPT_L2FILTER,
+    NETOPT_L2FILTER_RM, /**< get will always return -ENOTSUP */
```
Please add some type doc for this netopt, too.

miri64 replied:
Btw, I think you need to extend [`_netopt_strmap`](https://github.com/RIOT-OS/RIOT/blob/master/sys/net/crosslayer/netopt/netopt.c#L26-L62), too.

miri64 commented:
> I can confirm both blacklist and whitelist working on native and some boards. IMO this is a very useful tool which I will need in very near future. However, I'm not 100% happy with introducing two new NETOPT types to add and remove (will be done in a follow-up) entries. @miri64 what's your general opinion on this?

I'm not happy either, but for list operations this seems to be the only way at the momement.

smlng commented on sys/include/net/l2filter.h:
```diff
@@ -0,0 +1,113 @@
+ * @param[in] list      pointer to the filter list
+ * @param[in] addr      address to be added to list
+ * @param[in] addr_len  size of @p addr [in byte]
+ *
```
missing `@pre`s stating the assert conditions (list and addr != NULL, and addr_len < MAX_ADDRLEN.

smlng commented on sys/include/net/l2filter.h:
```diff
@@ -0,0 +1,113 @@
+ * @param[in] list      pointer to the filter list
+ * @param[in] addr      address to remove from the list
+ * @param[in] addr_len  length of @p addr [in byte]
+ *
```
missing `@pre`, see above

smlng commented on sys/include/net/l2filter.h:
```diff
@@ -0,0 +1,113 @@
+ * @param[in] list      list with black-/whitelisted addresses
+ * @param[in] addr      address to check against the entries in @p list
+ * @param[in] addr_len  length of @p addr [in byte]
+ *
```
dito see above regarding `@pre`

smlng commented on sys/include/net/l2filter.h:
```diff
@@ -0,0 +1,113 @@
+
+/**
+ * @brief   Number of slots in each filter list (filter entries per device)
+ */
```
inconsistent documentation here outside of `#ifndef`, above its within - don't whats (more) correct?

miri64 replied:
I prefer above, because then it is consistent with constructs like this

```C
/**
 * @brief Doc-brief
 * @note 1 if @ref sys_example is present, 2 if not.
 */
#ifdef MODULE_EXAMPLE
#define EXAMPLE_VALUE    (1)
#else
#define EXAMPLE_VALUE    (2)
#endif

miri64 replied:
(and yes there is a use-case for this construct. E.g. `gnrc_netif` uses it to calculate the number of required addresses).

smlng commented on sys/include/net/l2filter.h:
```diff
@@ -0,0 +1,113 @@
+/**
+ * @brief   Add an entry to a devices filter list
+ *
+ * @param[in] list      pointer to the filter list
```
this is actually `[in|out]`, isn't it?

PeterKietzmann replied:
It depends on the point of view but IMO its rather an `in` because you give **in** a pointer to the list where you want to add the entry.

smlng replied:
if it is an `in` than this could be`const`, but I doubt you can add an element then ...

cgundogan replied:
I am with @smlng here. The element that is pointed to by `list` (aka `list[0]`) will be modified (probably at the first addition)

haukepetersen replied:
yapp, but its `[in,out]` :-)

smlng commented on sys/include/net/l2filter.h:
```diff
@@ -0,0 +1,113 @@
+/**
+ * @brief   Remove an entry from the given filter list
+ *
+ * @param[in] list      pointer to the filter list
```
same as above: its `[in|out]` param

miri64 commented on sys/include/net/netopt.h:
```diff
@@ -248,6 +248,16 @@ typedef enum {
      * @note    Interfaces are not meant to respond to that.
      */
     NETOPT_NUMOF,
```
Uhm.... better put this below the newly defined `netopt`s ;-)

PeterKietzmann commented:
@miri64, @smlng I addressed your comments and PRed against @haukepetersen's branch [#39](https://github.com/haukepetersen/RIOT/pull/39). If it's just about the doc it would be so great if you approve so we can merge **soon**

cgundogan commented on sys/net/link_layer/l2filter/l2filter.c:
```diff
@@ -0,0 +1,106 @@
+        if (list[i].addr_len == 0) {
+            list[i].addr_len = addr_len;
+            memcpy(list[i].addr, addr, addr_len);
+            res = 0;
```
I would rather `break` here instead of testing `res` in the for loop header (all following for loops, too)

haukepetersen replied:
makes probably sense. Some old MISRA rules did not allow to use `break` (changed in newer MISRA rules, though), so I tend to be stuck not to use `break` when possible... But times change and I will try to adapt :-)

haukepetersen commented:
Tried to address all comments (@PeterKietzmann thanks for the help!)
- added `@pre`s -> is this convention now to do this for every function?
- fixed doc for new netopts
- use break in the implementation
- optimized shell command for add/removing filters

PeterKietzmann commented:
So I tested both list variants on boards and on native. Everything works as expected and as far as I see it, all comments were addressed. @miri64 , @smlng please approve

smlng commented on drivers/include/net/netdev.h:
```diff
@@ -54,6 +54,7 @@ extern "C" {
 #include <sys/uio.h>
 
 #include "net/netstats.h"
+#include "net/l2filter.h"
```
ifdef, too?

haukepetersen replied:
IMHO no need for this -> netstats does also not do this...

miri64 replied:
I'm also against `#ifdef` for header includes (unless 100% needed). Just bloats the top part of the code with more lines and the compiler ignores non-implemented header definitions anyway.

smlng replied:
yeah should be fine as long there is no global or external variable definition that could result in _undefined reference_ or _unused variable_ error when the module is not used.

cgundogan replied:
hm, thinking about @smlng's comment: including header files, where the implementation is known to be not present, may still lead to name clashes if those functions / variables are too generic in their names and in the worst case to undefined behavior (I assume there is no big #ifdef in the header file itself). IMO it's also better to take off some work for the preprocessor / compiler by hiding those includes, iff its known that they are not needed. But as @miri64 states, it could get ugly and bloated very quickly.

smlng replied:
so `#ifdef` in the header sounds like a good tradeoff here?

miri64 replied:
We do not use it in other modules where such constructs exist (e.g. `netstats.h`) so I wouldn't do it here either. If you are not happy with that, please provide a follow-up for these cases and let's discuss it there.

haukepetersen replied:
+1. Also: if there should be no name clashes in any way -> that's why we use the module name as prefix for everything a module defines...

cgundogan replied:
> If you are not happy with that, please provide a follow-up for these cases and let's discuss it there.

This is not how we **should** review our PRs. Rather than trying to get them in quickly, but sloppy, and rely on follow-up PRs, we should aim for our best in the first iteration.
Does it hurt so much to **not** copy a previously applied style that might be flawed? It's just two lines more for less problems .. Since when are unused include files not considered as bad coding? (;

miri64 replied:
*If* a clash occurs with a header from *our* codebase the header is flawed IMHO, not the consumer of that header (in the same codebase). I would rather take a look into the header and assure the clash does not happen there (e.g. with proper prefixing according to our naming conventions), and not in the consumers of that header. This is how you build a proper API: not putting the need to remember exceptions on the shoulders of the "user". Mind you, that is not true for external headers, but there you still have can name clashes in the linker (see emb6's clist e.g.) but those need to be fixed by other means anyway.

miri64 replied:
This is not sloppiness, this is moving tangential discussions (that stall the PR's review) to a separate thread BTW ;-).

cgundogan replied:
>If a clash occurs with a header from our codebase the header is flawed IMHO, not the consumer of that header (in the same codebase). I would rather take a look into the header and assure the clash does not happen there (e.g. with proper prefixing according to our naming conventions), and not in the consumers of that header.

Prefixing according to our coding conventions surely helps, but does not resolve the problem of clashing with external packages. You yourself gave a good example for what I am talking about.

Nevertheless, the question is not **how** to design a good API, but **whether** a header file should be included that is definitly not used. There are no arguments against `ifdefing` unused header files, besides the *bloating*. But, IMO, bloating is a good price to pay to reduce
1. global name space pollution
2. compilation time
3. strange behaviours due to compiler hints, like e.g. pragmas
4. potential clash of pre-processor statements
5. .. and depending on the order or a misconfigured build system, external software, where we do not have full control (besides of patching), might be affected by above bullet points.

I see that there is not really a point reasoning here as the discussion will spin around forever. I am just surprised that the harms of unused includes are not considered as a bad style and that flaws (IMO) are blindly copied.

@miri64 go ahead and merge.

miri64 replied:
Alright and thanks!

miri64 commented on makefiles/pseudomodules.inc.mk:
```diff
@@ -17,6 +17,7 @@ PSEUDOMODULES += gnrc_sixlowpan_router
 PSEUDOMODULES += gnrc_sixlowpan_router_default
 PSEUDOMODULES += gnrc_sock_check_reuse
 PSEUDOMODULES += gnrc_txtsnd
+PSEUDOMODULES += l2filter_%
```
So... `l2filter_foobar` gives me blacklist... Can you give explicit names here. Just having this translate to anything doesn't seem like a good idea. Pseudo-modules are already in-transparent as they are.

haukepetersen replied:
yes, will do

miri64 commented on sys/shell/commands/sc_netif.c:
```diff
@@ -143,6 +144,10 @@ static void _set_usage(char *cmd_name)
          "       * \"cca_threshold\" - set ED threshold during CCA in dBm\n"
+#ifdef MODULE_L2FILTER
+         "       * \"l2filter\" - add link layer address to filter\n"
+         "       * \"l2filter_rm\" - remove link layer address from filter\n"
```
Can the command be `ifconfig l2filter add` or `ifconfig l2filter del` please? IP addresses are also lists of addresses and are configured that way.

haukepetersen replied:
will also do

miri64 commented:
Tested for native. Apart from command structure and the pseudomodule names I'm okay with this PR

haukepetersen commented:
all fixed.

haukepetersen commented:
and one more minor thing that @miri64 pointed out offline.

haukepetersen commented:
squashed.

cgundogan commented:
>all fixed.

I know it's nice and less time-consuming to quickly merge open PRs, but how is https://github.com/RIOT-OS/RIOT/pull/7066#discussion_r117222118 considered as `fixed`? There was not even a discussion. IMO, unused includes are more `harmful` to RIOT than some people might think, especially if it's used throughout the code base.

haukepetersen commented:
@cgundogan: are you satisfied with the answers above? If yes, feel free to press the big green button :-)

miri64 commented:
>> all fixed.
>
> I know it's nice and less time-consuming to quickly merge open PRs, but how is #7066 (comment) considered as fixed? There was not even a discussion. IMO, unused includes are more harmful to RIOT than some people might think, especially if it's used throughout the code base.

@cgundogan As stated above that's a broader discussion that should be discussed in its own issue, so if you agree, please hit the merge button :-).

smlng commented:
@cgundogan: though I brought up the point and started this discussion in the first place, I'm with @miri64 here to solve this in a separate PR. Cause, as mentioned by @haukepetersen its the same for netstats (and maybe other modules, too). So fixing this here, does not resolve it for the rest. Hence, it make sense to create an issue on that matter or provide a PR for further discussion.
